### ISSUE TITLE
osi_linux: failed to read path.qstr_size

### ISSUE THREAD

**Issue by canance** [panda-re/panda#445]
Hi,

I'm trying to get `osi_test` to work on a Linux-based ARM image using the docker image of PANDA.  When I try to load the osi plugins using a replay, the `osi_linux` plugin reports that it fails to read `path.qstr_size`.  I believe that I have my kernelinfo.conf set up correctly and have attached it below.  Can anyone point me in the right direction for how to fix or debug this?

Thanks,

Cory

## Command line and monitor output
```
# qemu-system-arm -M versatilepb -kernel vmlinuz-3.2.0-4-versatile -initrd initrd.img-3.2.0-4-versatile -hda debian_wheezy_armel_standard.qcow2 -append "root=/dev/sda1 console=ttyAMA0" -vnc :1 -monitor stdio -replay take2 -panda "osi;osi_linux;osi_test" -os linux-32-debianarm
PANDA[core]:os_familyno=2 bits=32 os_details=debianarm
PANDA[core]:initializing osi
Looking for kconffile in /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf
OSI grabbing Linux introspection backend.
Linux OSI, using group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:E:kernelinfo_read.c(read_kernelinfo)> failed to read path.qstr_size
PANDA[osi_linux]:E:kernelinfo_read.c(read_kernelinfo)> 1 errors reading from group debianarm:32
PANDA[osi_linux]:W> kernelinfo bytes [76-79] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [200-207] not read
PANDA[osi_linux]:E:osi_linux.cpp(init_plugin)> Failed to read group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:FAILED to load required plugin osi_linux from /panda/build/arm-softmmu/panda/plugins/panda_osi_linux.so
Aborted (core dumped)
```

## kernelinfo.conf
```
[debianarm:32]
name = 3.2.0-4-versatile|#1 Debian 3.2.51-1|armv5tejl
version.a = 3
version.b = 2
version.c = 78
task.init_addr = 3224974096
#task.init_addr = 0xC0393310
task.size = 888
task.task_offset = 12
task.tasks_offset = 216
task.pid_offset = 260
task.tgid_offset = 264
task.group_leader_offset = 292
task.thread_group_offset = 348
task.real_parent_offset = 268
task.parent_offset = 272
task.mm_offset = 224
task.stack_offset = 4
task.real_cred_offset = 472
task.cred_offset = 476
task.comm_offset = 484
task.comm_size = 16
task.files_offset = 532
cred.uid_offset = 4
cred.gid_offset = 8
cred.euid_offset = 20
cred.egid_offset = 24
mm.size = 384
mm.mmap_offset = 0
mm.pgd_offset = 36
mm.arg_start_offset = 144
mm.start_brk_offset = 132
mm.brk_offset = 136
mm.start_stack_offset = 140
vma.size = 84
vma.vm_mm_offset = 0
vma.vm_start_offset = 4
vma.vm_end_offset = 8
vma.vm_next_offset = 12
vma.vm_flags_offset = 24
vma.vm_file_offset = 76
fs.f_path_dentry_offset = 12
fs.f_path_mnt_offset = 8
fs.f_pos_offset = 32
fs.fdt_offset = 4
fs.fdtab_offset = 8
fs.fd_offset = 4
qstr.size = 12
qstr.name_offset = 8
path.d_name_offset = 20
path.d_iname_offset = 36
path.d_parent_offset = 16
path.d_op_offset = 80
path.d_dname_offset = 28
path.mnt_root_offset = 16
path.mnt_parent_offset = 8
path.mnt_mountpoint_offset = 12

```

**Comment by m000** [comment]
Where did you get that kernel profile from? It seems that it uses an older format, where `qstr` is a top-level entry rather than nested under `path`. There have been some rearrangements to the kernel profile structs several months back.

Look at `kernelinfo.conf` in the repo. You may have some luck trying to map entries from the old profile to their new name. If the old profile misses some entries altogether, then unfortunately you'll have to extract a new kernel profile.

**Comment by nathanjackson** [comment]
We made some changes to `osi_linux` that made qstr a top-level entry when we added Linux 2.4.x support to `osi_linux`. We needed to do this to support the qstr struct layout changing.

This is the code on master that reads the qstr entries: https://github.com/panda-re/panda/blob/master/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo_read.c#L164

It seems like it should be working for you. Do you happen to be using an older version of PANDA than you have a kernelinfo.conf for?

**Comment by nathanjackson** [comment]
Oh, I just realized you're using the Docker image of PANDA.

I'm willing to bet the Docker image is out of date. I'm not sure who manages that. I'm betting it will work (or at least, get you past that point) by building the latest version.

**Comment by canance** [comment]
Hi @m000 and @nathanjackson,

Thank you both for your help!  You were right about the docker image being slightly behind the master.  I built a new docker image based on the current master and was able to get a little further :-).  

Now its failing to load `/panda/build/arm-softmmu/panda/plugins/panda_os=linux.so` (full monitor output below).  In that directory, there isn't a `panda_os=linux.so` file, but there is a `panda_osi_linux.so`.  Could it be meaning to load that .so instead?  

I copied `panda_osi_linux.so` to `panda_os=linux.so` and was able to get the replay to run successfully, but there was no output from the `osi_test` plugin.  I've also included the monitor output for this execution below.

Any further help would be greatly appreciated!

Thanks,
Cory  


### Monitor output (missing panda_os=linux.so)
```
root@iothoneypot:/app# ./run_replay.sh 
PANDA[core]:os_familyno=2 bits=32 os_details=debianarm
PANDA[core]:initializing osi
Looking for kconffile in /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf
OSI grabbing Linux introspection backend.
Linux OSI, using group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [108-111] not read
Failed to load /panda/build/arm-softmmu/panda/plugins/panda_os=linux.so: /panda/build/arm-softmmu/panda/plugins/panda_os=linux.so: cannot open shared object file: No such file or directory
FAIL: Unable to load plugin `/panda/build/arm-softmmu/panda/plugins/panda_os=linux.so'
./run_replay.sh: line 1:    34 Aborted                 (core dumped) qemu-system-arm -M versatilepb -kernel vmlinuz-3.2.0-4-versatile -initrd initrd.img-3.2.0-4-versatile -hda debian_wheezy_armel_standard.qcow2 -append "root=/dev/sda1 console=ttyAMA0" -vnc :1 -monitor stdio -replay take2 -panda "osi;os=linux;osi_linux;osi_test" -os linux-32-debianarm
```
### Monitor output (Missing osi_test output)
```
root@iothoneypot:/app# ./run_replay.sh 
PANDA[core]:os_familyno=2 bits=32 os_details=debianarm
PANDA[core]:initializing osi
Looking for kconffile in /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf
OSI grabbing Linux introspection backend.
Linux OSI, using group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [108-111] not read
PANDA[core]:initializing os=linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [108-111] not read
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi_linux.so already loaded
PANDA[core]:initializing osi_test
PANDA[core]:loading required plugin osi
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi.so already loaded
QEMU 2.9.1 monitor - type 'help' for more information
(qemu) audio: Could not init `oss' audio driver
loading snapshot
... done.
opening nondet log for read :	./take2-rr-nondet.log
./take2-rr-nondet.log:   840105469 instrs total.
take2:     8401056 (  1.00%) instrs.    0.72 sec.  0.11 GB ram.
take2:    16802129 (  2.00%) instrs.    0.82 sec.  0.11 GB ram.
take2:    25203176 (  3.00%) instrs.    0.87 sec.  0.11 GB ram.
take2:    33604221 (  4.00%) instrs.    0.92 sec.  0.11 GB ram.
take2:    42005274 (  5.00%) instrs.    0.97 sec.  0.11 GB ram.
take2:    50406343 (  6.00%) instrs.    1.04 sec.  0.11 GB ram.
take2:    58807383 (  7.00%) instrs.    1.10 sec.  0.11 GB ram.
take2:    67208447 (  8.00%) instrs.    1.14 sec.  0.11 GB ram.
take2:    75609511 (  9.00%) instrs.    1.20 sec.  0.11 GB ram.
take2:    84010571 ( 10.00%) instrs.    1.25 sec.  0.11 GB ram.
take2:    92411613 ( 11.00%) instrs.    1.29 sec.  0.11 GB ram.
take2:   100812677 ( 12.00%) instrs.    1.34 sec.  0.11 GB ram.
take2:   109213717 ( 13.00%) instrs.    1.38 sec.  0.11 GB ram.
take2:   117614770 ( 14.00%) instrs.    1.45 sec.  0.11 GB ram.
take2:   126015827 ( 15.00%) instrs.    1.51 sec.  0.11 GB ram.
take2:   134416898 ( 16.00%) instrs.    1.56 sec.  0.11 GB ram.
take2:   142817932 ( 17.00%) instrs.    1.60 sec.  0.11 GB ram.
take2:   151218991 ( 18.00%) instrs.    1.65 sec.  0.11 GB ram.
take2:   159620043 ( 19.00%) instrs.    1.70 sec.  0.11 GB ram.
take2:   168021112 ( 20.00%) instrs.    1.75 sec.  0.11 GB ram.
take2:   176422176 ( 21.00%) instrs.    1.80 sec.  0.11 GB ram.
take2:   184823204 ( 22.00%) instrs.    1.84 sec.  0.11 GB ram.
take2:   193224271 ( 23.00%) instrs.    1.90 sec.  0.11 GB ram.
take2:   201625326 ( 24.00%) instrs.    1.94 sec.  0.11 GB ram.
take2:   210026377 ( 25.00%) instrs.    1.99 sec.  0.11 GB ram.
take2:   218427431 ( 26.00%) instrs.    2.04 sec.  0.11 GB ram.
take2:   226828479 ( 27.00%) instrs.    2.12 sec.  0.11 GB ram.
take2:   235229532 ( 28.00%) instrs.    2.63 sec.  0.12 GB ram.
take2:   243630587 ( 29.00%) instrs.    2.96 sec.  0.13 GB ram.
take2:   252031642 ( 30.00%) instrs.    3.17 sec.  0.13 GB ram.
take2:   260432699 ( 31.00%) instrs.    3.38 sec.  0.13 GB ram.
take2:   268833751 ( 32.00%) instrs.    3.60 sec.  0.13 GB ram.
take2:   277234806 ( 33.00%) instrs.    3.95 sec.  0.13 GB ram.
take2:   285635862 ( 34.00%) instrs.    4.16 sec.  0.13 GB ram.
take2:   294036919 ( 35.00%) instrs.    4.36 sec.  0.13 GB ram.
take2:   302437969 ( 36.00%) instrs.    4.58 sec.  0.13 GB ram.
take2:   310839029 ( 37.00%) instrs.    4.80 sec.  0.13 GB ram.
take2:   319240079 ( 38.00%) instrs.    5.07 sec.  0.13 GB ram.
take2:   327641139 ( 39.00%) instrs.    5.39 sec.  0.13 GB ram.
take2:   336042189 ( 40.00%) instrs.    5.61 sec.  0.13 GB ram.
take2:   344443245 ( 41.00%) instrs.    5.82 sec.  0.13 GB ram.
take2:   352844308 ( 42.00%) instrs.    6.02 sec.  0.13 GB ram.
take2:   361245352 ( 43.00%) instrs.    6.22 sec.  0.13 GB ram.
take2:   369646410 ( 44.00%) instrs.    6.49 sec.  0.14 GB ram.
take2:   378047462 ( 45.00%) instrs.    6.91 sec.  0.14 GB ram.
take2:   386448519 ( 46.00%) instrs.    7.49 sec.  0.14 GB ram.
take2:   394849577 ( 47.00%) instrs.    7.69 sec.  0.14 GB ram.
take2:   403250636 ( 48.00%) instrs.    8.10 sec.  0.14 GB ram.
take2:   411651693 ( 49.00%) instrs.    8.47 sec.  0.15 GB ram.
take2:   420052736 ( 50.00%) instrs.    8.59 sec.  0.15 GB ram.
take2:   428453797 ( 51.00%) instrs.    8.73 sec.  0.15 GB ram.
take2:   436854844 ( 52.00%) instrs.    8.85 sec.  0.15 GB ram.
take2:   445255902 ( 53.00%) instrs.    9.00 sec.  0.15 GB ram.
take2:   453656958 ( 54.00%) instrs.    9.19 sec.  0.15 GB ram.
take2:   462058010 ( 55.00%) instrs.    9.36 sec.  0.15 GB ram.
take2:   470459064 ( 56.00%) instrs.    9.49 sec.  0.15 GB ram.
take2:   478860122 ( 57.00%) instrs.    9.62 sec.  0.15 GB ram.
take2:   487261175 ( 58.00%) instrs.    9.81 sec.  0.15 GB ram.
take2:   495662234 ( 59.00%) instrs.    9.94 sec.  0.15 GB ram.
take2:   504063287 ( 60.00%) instrs.   10.06 sec.  0.15 GB ram.
take2:   512464340 ( 61.00%) instrs.   10.18 sec.  0.15 GB ram.
take2:   520865400 ( 62.00%) instrs.   10.30 sec.  0.15 GB ram.
take2:   529266453 ( 63.00%) instrs.   10.43 sec.  0.15 GB ram.
take2:   537667501 ( 64.00%) instrs.   10.55 sec.  0.15 GB ram.
take2:   546068566 ( 65.00%) instrs.   10.68 sec.  0.15 GB ram.
take2:   554469620 ( 66.00%) instrs.   10.80 sec.  0.15 GB ram.
take2:   562870667 ( 67.00%) instrs.   10.92 sec.  0.15 GB ram.
take2:   571271730 ( 68.00%) instrs.   11.03 sec.  0.15 GB ram.
take2:   579672783 ( 69.00%) instrs.   11.17 sec.  0.15 GB ram.
take2:   588073833 ( 70.00%) instrs.   11.29 sec.  0.15 GB ram.
take2:   596474896 ( 71.00%) instrs.   11.41 sec.  0.15 GB ram.
take2:   604875951 ( 72.00%) instrs.   11.54 sec.  0.15 GB ram.
take2:   613276998 ( 73.00%) instrs.   11.65 sec.  0.15 GB ram.
take2:   621678056 ( 74.00%) instrs.   11.76 sec.  0.15 GB ram.
take2:   630079115 ( 75.00%) instrs.   11.87 sec.  0.15 GB ram.
take2:   638480168 ( 76.00%) instrs.   11.98 sec.  0.15 GB ram.
take2:   646881217 ( 77.00%) instrs.   12.10 sec.  0.15 GB ram.
take2:   655282269 ( 78.00%) instrs.   12.21 sec.  0.15 GB ram.
take2:   663683321 ( 79.00%) instrs.   12.34 sec.  0.15 GB ram.
take2:   672084381 ( 80.00%) instrs.   12.46 sec.  0.15 GB ram.
take2:   680485432 ( 81.00%) instrs.   12.57 sec.  0.15 GB ram.
take2:   688886493 ( 82.00%) instrs.   12.69 sec.  0.15 GB ram.
take2:   697287541 ( 83.00%) instrs.   12.80 sec.  0.15 GB ram.
take2:   705688603 ( 84.00%) instrs.   12.94 sec.  0.15 GB ram.
take2:   714089649 ( 85.00%) instrs.   13.10 sec.  0.15 GB ram.
take2:   722490706 ( 86.00%) instrs.   13.22 sec.  0.15 GB ram.
take2:   730891760 ( 87.00%) instrs.   13.34 sec.  0.15 GB ram.
take2:   739292814 ( 88.00%) instrs.   13.46 sec.  0.15 GB ram.
take2:   747693881 ( 89.00%) instrs.   13.59 sec.  0.15 GB ram.
take2:   756094935 ( 90.00%) instrs.   13.72 sec.  0.15 GB ram.
take2:   764495987 ( 91.00%) instrs.   13.89 sec.  0.15 GB ram.
take2:   772897036 ( 92.00%) instrs.   14.00 sec.  0.15 GB ram.
take2:   781298093 ( 93.00%) instrs.   14.12 sec.  0.15 GB ram.
take2:   789699151 ( 94.00%) instrs.   14.26 sec.  0.15 GB ram.
take2:   798100197 ( 95.00%) instrs.   14.38 sec.  0.15 GB ram.
take2:   806501257 ( 96.00%) instrs.   14.50 sec.  0.15 GB ram.
take2:   814902313 ( 97.00%) instrs.   14.62 sec.  0.15 GB ram.
take2:   823303362 ( 98.00%) instrs.   14.75 sec.  0.15 GB ram.
take2:   831704428 ( 99.00%) instrs.   14.90 sec.  0.15 GB ram.
./take2-rr-nondet.log:  log is empty.
./take2-rr-nondet.log:  log is empty.
Replay completed successfully. 1
Time taken was: 15 seconds.
Stats:
RR_INPUT_1 number = 0, size = 0 bytes
RR_INPUT_2 number = 0, size = 0 bytes
RR_INPUT_4 number = 0, size = 0 bytes
RR_INPUT_8 number = 24261, size = 436698 bytes
RR_INTERRUPT_REQUEST number = 41277, size = 577878 bytes
RR_EXIT_REQUEST number = 0, size = 0 bytes
RR_SKIPPED_CALL number = 2030, size = 1550037 bytes
RR_END_OF_LOG number = 1, size = 10 bytes
RR_PENDING_INTERRUPTS number = 0, size = 0 bytes
RR_EXCEPTION number = 0, size = 0 bytes
max_queue_len = 772
Checksum of guest memory: 0xbbeee37d
Replay completed successfully 2.
```

**Comment by m000** [comment]
The command line used by the wrapper script doesn't look right: `-panda "osi;os=linux;osi_linux;osi_test"`

The `os=linux` part shouldn't be there at all. `-os ...` is a top level option, not a plugin argument.

**Comment by canance** [comment]
Thanks @m000!  That was a mistake on my part.  I've removed that and retried the `osi_test` module with the following command line:
`qemu-system-arm -M versatilepb -kernel vmlinuz-3.2.0-4-versatile -initrd initrd.img-3.2.0-4-versatile -hda debian_wheezy_armel_standard.qcow2 -append "root=/dev/sda1 console=ttyAMA0" -vnc :1 -monitor stdio -replay recording1 -panda "osi;osi_linux;osi_test" -os linux-32-debianarm`

I can see that `osi_test` is initialized and it finds the correct kconf_group, but the plugin doesn't produce any output.  I've provided the output below.

PS -- my recording is of `wget` grabbing an html page and then `cat`'ing the contents.

Thanks!
Cory

### Output
```
PANDA[core]:os_familyno=2 bits=32 os_details=debianarm
PANDA[core]:initializing osi
Looking for kconffile in /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf
OSI grabbing Linux introspection backend.
Linux OSI, using group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [108-111] not read
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi_linux.so already loaded
PANDA[core]:initializing osi_test
PANDA[core]:loading required plugin osi
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi.so already loaded
QEMU 2.9.1 monitor - type 'help' for more information
(qemu) audio: Could not init `oss' audio driver
loading snapshot
... done.
opening nondet log for read :	./recording1-rr-nondet.log
./recording1-rr-nondet.log:  1972977469 instrs total.
recording1:    19729779 (  1.00%) instrs.    1.11 sec.  0.11 GB ram.
[...]
recording1:  1953247695 ( 99.00%) instrs.   54.45 sec.  0.15 GB ram.
./recording1-rr-nondet.log:  log is empty.
./recording1-rr-nondet.log:  log is empty.
Replay completed successfully. 1
Time taken was: 55 seconds.
Stats:
RR_INPUT_1 number = 0, size = 0 bytes
RR_INPUT_2 number = 0, size = 0 bytes
RR_INPUT_4 number = 0, size = 0 bytes
RR_INPUT_8 number = 29432, size = 529776 bytes
RR_INTERRUPT_REQUEST number = 43863, size = 614082 bytes
RR_EXIT_REQUEST number = 0, size = 0 bytes
RR_SKIPPED_CALL number = 5683, size = 3005265 bytes
RR_END_OF_LOG number = 1, size = 10 bytes
RR_PENDING_INTERRUPTS number = 0, size = 0 bytes
RR_EXCEPTION number = 0, size = 0 bytes
max_queue_len = 1501
Checksum of guest memory: 0x75452dcf
Replay completed successfully 2.
```

**Comment by nathanjackson** [comment]
I've honestly never tried ARM guests so I have no idea what the status of it is.

Can you try the `asidstory` plugin? I'm curious if it works or if there is a bigger problem with OSI and ARM.

**Comment by canance** [comment]
Sure, no problem.

It appears to not work as well.  It errors out in `common.c:30:regime_el` (full output below).  Looking at the source, it's hitting the default case in the switch statement:
```
#ifdef TARGET_ARM
/* Return the exception level which controls this address translation regime */
static inline uint32_t regime_el(CPUARMState *env, ARMMMUIdx mmu_idx)
{
    switch (mmu_idx) {
    case ARMMMUIdx_S2NS:
    case ARMMMUIdx_S1E2:
        return 2;
    case ARMMMUIdx_S1E3:
        return 3;
    case ARMMMUIdx_S1SE0:
        return arm_el_is_aa64(env, 3) ? 1 : 3;
    case ARMMMUIdx_S1SE1:
    case ARMMMUIdx_S1NSE0:
    case ARMMMUIdx_S1NSE1:
    case ARMMMUIdx_S12NSE1:
        return 1;
    default:
        g_assert_not_reached();
    }
}
```

I would if the `versatilepb` machine that I'm emulating doesn't have a TLB?  A quick search doesn't turn up much information about that.

# Output
```
# qemu-system-arm -M versatilepb -kernel vmlinuz-3.2.0-4-versatile -initrd initrd.img-3.2.0-4-versatile -hda debian_wheezy_armel_standard.qcow2 -append "root=/dev/sda1 console=ttyAMA0" -vnc :1 -monitor stdio -replay recording1 -panda "osi;osi_linux;asidstory:width=180" -os linux-32-debianarm
PANDA[asidstory]:adding argument width=180.
PANDA[core]:os_familyno=2 bits=32 os_details=debianarm
PANDA[core]:initializing osi
Looking for kconffile in /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf
OSI grabbing Linux introspection backend.
Linux OSI, using group debianarm:32 from /panda/build/arm-softmmu/panda/plugins/osi_linux/kernelinfo.conf.
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [92-95] not read
PANDA[osi_linux]:W> kernelinfo bytes [108-111] not read
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi_linux.so already loaded
PANDA[core]:initializing asidstory
PANDA[core]:loading required plugin osi
PANDA[core]:/panda/build/arm-softmmu/panda/plugins/panda_osi.so already loaded
QEMU 2.9.1 monitor - type 'help' for more information
(qemu) audio: Could not init `oss' audio driver
loading snapshot
... done.
opening nondet log for read :	./recording1-rr-nondet.log
**
ERROR:/panda/panda/src/common.c:30:regime_el: code should not be reached
```

**Comment by AndrewFasano** [comment]
I believe I fixed this issue in 048be77348b08606eabe3290604721a6eccdc940.

We're currently trying to add some (previously missing) stability to using PANDA with ARM. After #620 gets merged, I believe OSI support should work reasonably well on ARM.
